File: tools/memory/memory_quality_gate.py
#!/usr/bin/env python3
from __future__ import annotations
import argparse, json, os, re, subprocess
from dataclasses import dataclass, asdict
from datetime import datetime, timezone
from difflib import SequenceMatcher
from typing import Any

DB_BIN='/opt/homebrew/opt/postgresql@17/bin'; DB='cortana'
ACTION={'must','should','need','todo','remember','prefer','rule','plan','schedule','review','fix','send','build'}
LONG={'always','never','preference','policy','rule','habit','routine','goal','weekly','monthly','career','health','finance'}
SHORT={'today','tomorrow','asap','immediately','now','temporary','one-time'}
NEG={' not ',' never ',' no ',' don\'t ',' do not ',' avoid ',' stop '}

@dataclass
class Candidate:
    text:str; source_type:str='manual'; source_ref:str|None=None; timestamp:str|None=None

def run_psql(sql:str)->str:
    env=os.environ.copy(); env['PATH']=f"{DB_BIN}:{env.get('PATH','')}"
    p=subprocess.run(['psql',DB,'-q','-X','-v','ON_ERROR_STOP=1','-t','-A','-c',sql],capture_output=True,text=True,env=env)
    if p.returncode!=0: raise RuntimeError(p.stderr.strip() or 'psql failed')
    return p.stdout.strip()

def fetch_json(sql:str)->list[dict[str,Any]]:
    raw=run_psql(f"SELECT COALESCE(json_agg(t),'[]'::json)::text FROM ({sql}) t;")
    return json.loads(raw) if raw else []

def toks(t:str)->set[str]: return set(re.findall(r"[A-Za-z][A-Za-z0-9_'-]{2,}",t.lower()))

def sim(a:str,b:str)->float:
    if not a or not b: return 0.0
    seq=SequenceMatcher(None,a.lower(),b.lower()).ratio(); ta,tb=toks(a),toks(b)
    jac=(len(ta&tb)/len(ta|tb)) if (ta or tb) else 0.0
    return max(seq,jac)

def corpus(limit:int)->list[dict[str,Any]]:
    sql=f"""
    WITH e AS (SELECT 'episodic' tier,id::text id,COALESCE(happened_at::text,NOW()::text) ts,TRIM(COALESCE(summary,'')||' '||COALESCE(details,'')) body FROM cortana_memory_episodic WHERE active=TRUE ORDER BY happened_at DESC NULLS LAST LIMIT {limit}),
         s AS (SELECT 'semantic' tier,id::text id,COALESCE(last_seen_at::text,first_seen_at::text,NOW()::text) ts,TRIM(COALESCE(subject,'')||' '||COALESCE(predicate,'')||' '||COALESCE(object_value,'')) body FROM cortana_memory_semantic WHERE active=TRUE ORDER BY last_seen_at DESC NULLS LAST LIMIT {limit}),
         p AS (SELECT 'procedural' tier,id::text id,COALESCE(updated_at::text,created_at::text,NOW()::text) ts,TRIM(COALESCE(workflow_name,'')||' '||COALESCE(trigger_context,'')||' '||COALESCE(expected_outcome,'')) body FROM cortana_memory_procedural WHERE deprecated=FALSE ORDER BY updated_at DESC NULLS LAST LIMIT {limit})
    SELECT * FROM (SELECT * FROM e UNION ALL SELECT * FROM s UNION ALL SELECT * FROM p) x WHERE COALESCE(body,'')<>'' ORDER BY ts DESC LIMIT {limit*2}
    """
    try:
        return fetch_json(sql)
    except Exception:
        return []

def score_actionability(text:str)->float:
    t=f" {text.lower()} "; hits=sum(1 for w in ACTION if (w in toks(t) or f" {w} " in t))
    has_verb=bool(re.search(r"\b(do|build|send|review|fix|create|schedule|call|ship|track|improve)\b",t))
    has_time=bool(re.search(r"\b(today|tomorrow|week|month|by\s+\w+day|\d{1,2}:\d{2})\b",t))
    return round(min(1.0,0.25+0.12*hits+(0.15 if has_verb else 0)+(0.15 if has_time else 0)),3)

def score_shelf(text:str)->float:
    t=text.lower(); long=sum(1 for w in LONG if w in t); short=sum(1 for w in SHORT if w in t)
    if long==0 and short==0: return 0.55
    return round(max(0,min(1,0.5+0.1*long-0.12*short)),3)

def contradiction(new:str,old:str)->tuple[bool,str]:
    ov=toks(new)&toks(old)
    if len(ov)<3: return False,'low overlap'
    nneg=any(w in f" {new.lower()} " for w in NEG); oneg=any(w in f" {old.lower()} " for w in NEG)
    if nneg!=oneg: return True,'polarity flip with shared anchors'
    return False,'none'

def evaluate(c:Candidate,limit:int=300)->dict[str,Any]:
    mem=corpus(limit); scored=[]
    for r in mem:
        s=sim(c.text,str(r.get('body') or ''))
        if s>=0.35: scored.append((s,r))
    scored.sort(key=lambda x:x[0],reverse=True); top=scored[:8]
    rec=len([1 for s,_ in scored if s>=0.72]); novelty=round(max(0,1-(top[0][0] if top else 0)),3)
    recurrence=round(max(0,1-min(rec,8)/8),3); action=score_actionability(c.text); shelf=score_shelf(c.text)
    weighted=round(novelty*0.35+action*0.25+recurrence*0.2+shelf*0.2,3)
    verdict='promote' if (weighted>=0.68 and novelty>=0.4) else ('hold' if weighted>=0.45 else 'archive')
    supers=[]; now=c.timestamp or datetime.now(timezone.utc).isoformat()
    for s,r in top:
        ok,why=contradiction(c.text,str(r.get('body') or ''))
        if ok: supers.append({'tier':r.get('tier'),'id':r.get('id'),'similarity':round(s,3),'demote_recommended':True,'reason':why,'new_memory_timestamp':now})
    reasons=[]
    if novelty<0.35: reasons.append('low novelty')
    if action<0.35: reasons.append('low actionability')
    if shelf<0.35: reasons.append('short shelf-life')
    return {
      'verdict':verdict,
      'scores':{'novelty':novelty,'actionability':action,'recurrence':recurrence,'shelf_life':shelf,'weighted':weighted},
      'recurrence_count':rec,
      'matched_examples':[{'tier':r.get('tier'),'id':r.get('id'),'similarity':round(s,3),'text_preview':str(r.get('body') or '')[:180]} for s,r in top],
      'supersession_flags':supers,
      'reasons':reasons,
    }

def log_event(c:Candidate,res:dict[str,Any]):
    payload=json.dumps({'memory_text':c.text[:500],'source_type':c.source_type,'source_ref':c.source_ref,'result':res}).replace("'","''")
    msg=(f"Evaluated memory candidate: {res['verdict']}").replace("'","''")
    run_psql(f"INSERT INTO cortana_events (event_type,source,severity,message,metadata) VALUES ('memory_quality_gate','memory_quality_gate.py','info','{msg}','{payload}'::jsonb);")

def main()->int:
    ap=argparse.ArgumentParser(description='Score memory quality and return consolidation verdict')
    ap.add_argument('--text'); ap.add_argument('--text-file'); ap.add_argument('--source-type',default='manual')
    ap.add_argument('--source-ref'); ap.add_argument('--timestamp'); ap.add_argument('--corpus-limit',type=int,default=300)
    ap.add_argument('--log-event',action='store_true'); ap.add_argument('--dry-run',action='store_true')
    a=ap.parse_args(); text=a.text or ''
    if a.text_file: text=open(a.text_file,'r',encoding='utf-8').read().strip()
    if not text: raise SystemExit('Provide --text or --text-file')
    c=Candidate(text=text,source_type=a.source_type,source_ref=a.source_ref,timestamp=a.timestamp)
    res=evaluate(c,a.corpus_limit)
    if a.log_event and not a.dry_run: log_event(c,res)
    print(json.dumps(res,indent=2)); return 0

if __name__=='__main__': raise SystemExit(main())

File: tools/memory/memory_quality_gate.ts
#!/usr/bin/env npx tsx

import fs from "fs";
import { query } from "../lib/db.js";

const ACTION = new Set([
  "must",
  "should",
  "need",
  "todo",
  "remember",
  "prefer",
  "rule",
  "plan",
  "schedule",
  "review",
  "fix",
  "send",
  "build",
]);
const LONG = new Set([
  "always",
  "never",
  "preference",
  "policy",
  "rule",
  "habit",
  "routine",
  "goal",
  "weekly",
  "monthly",
  "career",
  "health",
  "finance",
]);
const SHORT = new Set(["today", "tomorrow", "asap", "immediately", "now", "temporary", "one-time"]);
const NEG = [" not ", " never ", " no ", " don't ", " do not ", " avoid ", " stop "];

type Candidate = {
  text: string;
  source_type: string;
  source_ref: string | null;
  timestamp: string | null;
};

type MatchBlock = { i: number; j: number; size: number };

function runPsql(sql: string): string {
  return query(sql).trim();
}

function fetchJson(sql: string): Array<Record<string, any>> {
  const raw = runPsql(`SELECT COALESCE(json_agg(t),'[]'::json)::text FROM (${sql}) t;`);
  return raw ? (JSON.parse(raw) as Array<Record<string, any>>) : [];
}

function toks(t: string): Set<string> {
  const matches = t.match(/[A-Za-z][A-Za-z0-9_'-]{2,}/g) ?? [];
  return new Set(matches.map((m) => m.toLowerCase()));
}

function buildB2j(b: string[], autojunk = true): Map<string, number[]> {
  const b2j = new Map<string, number[]>();
  for (let i = 0; i < b.length; i += 1) {
    const elt = b[i];
    const arr = b2j.get(elt);
    if (arr) arr.push(i);
    else b2j.set(elt, [i]);
  }
  if (autojunk && b.length >= 200) {
    const ntest = Math.floor(b.length / 100) + 1;
    for (const [elt, idxs] of b2j.entries()) {
      if (idxs.length > ntest) {
        b2j.delete(elt);
      }
    }
  }
  return b2j;
}

function findLongestMatch(
  a: string[],
  b: string[],
  alo: number,
  ahi: number,
  blo: number,
  bhi: number,
  b2j: Map<string, number[]>
): MatchBlock {
  let besti = alo;
  let bestj = blo;
  let bestsize = 0;
  let j2len = new Map<number, number>();

  for (let i = alo; i < ahi; i += 1) {
    const newj2len = new Map<number, number>();
    const indices = b2j.get(a[i]) ?? [];
    for (const j of indices) {
      if (j < blo) continue;
      if (j >= bhi) break;
      const k = (j2len.get(j - 1) ?? 0) + 1;
      newj2len.set(j, k);
      if (k > bestsize) {
        besti = i - k + 1;
        bestj = j - k + 1;
        bestsize = k;
      }
    }
    j2len = newj2len;
  }

  return { i: besti, j: bestj, size: bestsize };
}

function getMatchingBlocks(a: string[], b: string[]): MatchBlock[] {
  const la = a.length;
  const lb = b.length;
  const b2j = buildB2j(b);
  const queue: Array<[number, number, number, number]> = [[0, la, 0, lb]];
  const matching: MatchBlock[] = [];

  while (queue.length) {
    const [alo, ahi, blo, bhi] = queue.pop() as [number, number, number, number];
    const match = findLongestMatch(a, b, alo, ahi, blo, bhi, b2j);
    if (match.size) {
      if (alo < match.i && blo < match.j) {
        queue.push([alo, match.i, blo, match.j]);
      }
      if (match.i + match.size < ahi && match.j + match.size < bhi) {
        queue.push([match.i + match.size, ahi, match.j + match.size, bhi]);
      }
      matching.push(match);
    }
  }

  matching.sort((x, y) => (x.i - y.i) || (x.j - y.j));
  const nonAdjacent: MatchBlock[] = [];
  let i1 = 0;
  let j1 = 0;
  let k1 = 0;
  for (const m of matching) {
    if (m.i === i1 + k1 && m.j === j1 + k1) {
      k1 += m.size;
    } else {
      if (k1) nonAdjacent.push({ i: i1, j: j1, size: k1 });
      i1 = m.i;
      j1 = m.j;
      k1 = m.size;
    }
  }
  if (k1) nonAdjacent.push({ i: i1, j: j1, size: k1 });
  nonAdjacent.push({ i: la, j: lb, size: 0 });
  return nonAdjacent;
}

function sequenceMatcherRatio(a: string, b: string): number {
  if (!a && !b) return 1.0;
  if (!a || !b) return 0.0;
  const aa = a.split("");
  const bb = b.split("");
  const blocks = getMatchingBlocks(aa, bb);
  const matches = blocks.reduce((sum, blk) => sum + blk.size, 0);
  return (2.0 * matches) / (aa.length + bb.length);
}

function sim(a: string, b: string): number {
  if (!a || !b) return 0.0;
  const seq = sequenceMatcherRatio(a.toLowerCase(), b.toLowerCase());
  const ta = toks(a);
  const tb = toks(b);
  const inter = new Set([...ta].filter((x) => tb.has(x)));
  const union = new Set([...ta, ...tb]);
  const jac = union.size ? inter.size / union.size : 0.0;
  return Math.max(seq, jac);
}

function corpus(limit: number): Array<Record<string, any>> {
  const sql = `
    WITH e AS (SELECT 'episodic' tier,id::text id,COALESCE(happened_at::text,NOW()::text) ts,TRIM(COALESCE(summary,'')||' '||COALESCE(details,'')) body FROM cortana_memory_episodic WHERE active=TRUE ORDER BY happened_at DESC NULLS LAST LIMIT ${limit}),
         s AS (SELECT 'semantic' tier,id::text id,COALESCE(last_seen_at::text,first_seen_at::text,NOW()::text) ts,TRIM(COALESCE(subject,'')||' '||COALESCE(predicate,'')||' '||COALESCE(object_value,'')) body FROM cortana_memory_semantic WHERE active=TRUE ORDER BY last_seen_at DESC NULLS LAST LIMIT ${limit}),
         p AS (SELECT 'procedural' tier,id::text id,COALESCE(updated_at::text,created_at::text,NOW()::text) ts,TRIM(COALESCE(workflow_name,'')||' '||COALESCE(trigger_context,'')||' '||COALESCE(expected_outcome,'')) body FROM cortana_memory_procedural WHERE deprecated=FALSE ORDER BY updated_at DESC NULLS LAST LIMIT ${limit})
    SELECT * FROM (SELECT * FROM e UNION ALL SELECT * FROM s UNION ALL SELECT * FROM p) x WHERE COALESCE(body,'')<>'' ORDER BY ts DESC LIMIT ${limit * 2}
  `;
  try {
    return fetchJson(sql);
  } catch {
    return [];
  }
}

function scoreActionability(text: string): number {
  const t = ` ${text.toLowerCase()} `;
  const tokens = toks(t);
  let hits = 0;
  for (const w of ACTION) {
    if (tokens.has(w) || t.includes(` ${w} `)) hits += 1;
  }
  const hasVerb = /\b(do|build|send|review|fix|create|schedule|call|ship|track|improve)\b/.test(t);
  const hasTime = /\b(today|tomorrow|week|month|by\s+\w+day|\d{1,2}:\d{2})\b/.test(t);
  const score = 0.25 + 0.12 * hits + (hasVerb ? 0.15 : 0) + (hasTime ? 0.15 : 0);
  return Math.round(Math.min(1.0, score) * 1000) / 1000;
}

function scoreShelf(text: string): number {
  const t = text.toLowerCase();
  let long = 0;
  let short = 0;
  for (const w of LONG) if (t.includes(w)) long += 1;
  for (const w of SHORT) if (t.includes(w)) short += 1;
  if (long === 0 && short === 0) return 0.55;
  const score = 0.5 + 0.1 * long - 0.12 * short;
  return Math.round(Math.max(0, Math.min(1, score)) * 1000) / 1000;
}

function contradiction(neu: string, old: string): [boolean, string] {
  const ov = new Set([...toks(neu)].filter((x) => toks(old).has(x)));
  if (ov.size < 3) return [false, "low overlap"];
  const nneg = NEG.some((w) => ` ${neu.toLowerCase()} `.includes(w));
  const oneg = NEG.some((w) => ` ${old.toLowerCase()} `.includes(w));
  if (nneg !== oneg) return [true, "polarity flip with shared anchors"];
  return [false, "none"];
}

function evaluate(c: Candidate, limit = 300): Record<string, any> {
  const mem = corpus(limit);
  const scored: Array<[number, Record<string, any>]> = [];
  for (const r of mem) {
    const s = sim(c.text, String(r.body ?? ""));
    if (s >= 0.35) scored.push([s, r]);
  }
  scored.sort((a, b) => b[0] - a[0]);
  const top = scored.slice(0, 8);
  const rec = scored.filter((x) => x[0] >= 0.72).length;
  const novelty = Math.round(Math.max(0, 1 - (top.length ? top[0][0] : 0)) * 1000) / 1000;
  const recurrence = Math.round(Math.max(0, 1 - Math.min(rec, 8) / 8) * 1000) / 1000;
  const action = scoreActionability(c.text);
  const shelf = scoreShelf(c.text);
  const weighted = Math.round((novelty * 0.35 + action * 0.25 + recurrence * 0.2 + shelf * 0.2) * 1000) / 1000;
  const verdict = weighted >= 0.68 && novelty >= 0.4 ? "promote" : weighted >= 0.45 ? "hold" : "archive";
  const supers: Array<Record<string, any>> = [];
  const now = c.timestamp ?? new Date().toISOString();
  for (const [s, r] of top) {
    const [ok, why] = contradiction(c.text, String(r.body ?? ""));
    if (ok) {
      supers.push({
        tier: r.tier,
        id: r.id,
        similarity: Math.round(s * 1000) / 1000,
        demote_recommended: true,
        reason: why,
        new_memory_timestamp: now,
      });
    }
  }
  const reasons: string[] = [];
  if (novelty < 0.35) reasons.push("low novelty");
  if (action < 0.35) reasons.push("low actionability");
  if (shelf < 0.35) reasons.push("short shelf-life");
  return {
    verdict,
    scores: { novelty, actionability: action, recurrence, shelf_life: shelf, weighted },
    recurrence_count: rec,
    matched_examples: top.map(([s, r]) => ({
      tier: r.tier,
      id: r.id,
      similarity: Math.round(s * 1000) / 1000,
      text_preview: String(r.body ?? "").slice(0, 180),
    })),
    supersession_flags: supers,
    reasons,
  };
}

function sqlEscape(text: string): string {
  return text.replace(/'/g, "''");
}

function logEvent(c: Candidate, res: Record<string, any>): void {
  const payload = JSON.stringify({
    memory_text: c.text.slice(0, 500),
    source_type: c.source_type,
    source_ref: c.source_ref,
    result: res,
  }).replace(/'/g, "''");
  const msg = `Evaluated memory candidate: ${res.verdict}`.replace(/'/g, "''");
  runPsql(
    `INSERT INTO cortana_events (event_type,source,severity,message,metadata) VALUES ('memory_quality_gate','memory_quality_gate.ts','info','${msg}','${payload}'::jsonb);`
  );
}

function printHelp(): void {
  const text = `usage: memory_quality_gate.ts [-h] [--text TEXT] [--text-file TEXT_FILE] [--source-type SOURCE_TYPE] [--source-ref SOURCE_REF] [--timestamp TIMESTAMP] [--corpus-limit CORPUS_LIMIT] [--log-event] [--dry-run]\n\nScore memory quality and return consolidation verdict\n\noptions:\n  -h, --help            show this help message and exit\n  --text TEXT\n  --text-file TEXT_FILE\n  --source-type SOURCE_TYPE\n  --source-ref SOURCE_REF\n  --timestamp TIMESTAMP\n  --corpus-limit CORPUS_LIMIT\n  --log-event\n  --dry-run`;
  console.log(text);
}

async function main(): Promise<void> {
  const argv = process.argv.slice(2);
  let text = "";
  let textFile: string | null = null;
  let sourceType = "manual";
  let sourceRef: string | null = null;
  let timestamp: string | null = null;
  let corpusLimit = 300;
  let logEventFlag = false;
  let dryRun = false;

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = argv[i + 1];
    if (arg === "-h" || arg === "--help") {
      printHelp();
      process.exit(0);
    } else if (arg === "--text" && next) {
      text = next;
      i += 1;
    } else if (arg === "--text-file" && next) {
      textFile = next;
      i += 1;
    } else if (arg === "--source-type" && next) {
      sourceType = next;
      i += 1;
    } else if (arg === "--source-ref" && next) {
      sourceRef = next;
      i += 1;
    } else if (arg === "--timestamp" && next) {
      timestamp = next;
      i += 1;
    } else if (arg === "--corpus-limit" && next) {
      corpusLimit = Number.parseInt(next, 10);
      i += 1;
    } else if (arg === "--log-event") {
      logEventFlag = true;
    } else if (arg === "--dry-run") {
      dryRun = true;
    } else if (arg.startsWith("-")) {
      console.error(`Unknown argument: ${arg}`);
      printHelp();
      process.exit(2);
    }
  }

  if (textFile) {
    text = fs.readFileSync(textFile, "utf8").trim();
  }

  if (!text) {
    console.error("Provide --text or --text-file");
    process.exit(1);
  }

  const c: Candidate = { text, source_type: sourceType, source_ref: sourceRef, timestamp };
  const res = evaluate(c, corpusLimit);
  if (logEventFlag && !dryRun) logEvent(c, res);
  console.log(JSON.stringify(res, null, 2));
}

main().catch((err) => {
  const msg = err instanceof Error ? err.message : String(err);
  console.error(msg);
  process.exit(1);
});
